src/storage/gc_worker.rs

```diff
@@ -0,0 +1,305 @@
+            // TODO: Duplicated code in scheduler.rs
+            let gc_info = txn.gc(&k, safe_point)?;
+
+            if gc_info.found_versions >= GC_LOG_FOUND_VERSION_THRESHOLD {
+                info!(
+                    "[region {}] GC found at least {} versions for key {}",
+                    ctx.get_region_id(),
+                    gc_info.found_versions,
+                    k
+                );
+            }
+            // TODO: we may delete only part of the versions in a batch, which may not beyond
+            // the logging threshold `GC_LOG_DELETED_VERSION_THRESHOLD`.
+            if gc_info.deleted_versions as usize >= GC_LOG_DELETED_VERSION_THRESHOLD {
+                info!(
+                    "[region {}] GC deleted {} versions for key {}",
+                    ctx.get_region_id(),
+                    gc_info.deleted_versions,
+                    k
+                );
+            }
+
+            if txn.write_size() >= MAX_TXN_WRITE_SIZE {
+                next_scan_key = Some(k);
+                break;
+            }
+        }
+        self.stats.add_statistics(txn.get_statistics());
+
+        let modifies = txn.into_modifies();
+        if !modifies.is_empty() {
+            self.engine.write(ctx, modifies)?;
+        }
+        Ok(next_scan_key)
+    }
+
+    pub fn gc(&mut self, mut ctx: Context, safe_point: u64) -> Result<()> {
+        debug!(
+            "doing gc on region {}, safe_point {}",
+            ctx.get_region_id(),
+            safe_point
+        );
+
+        let mut next_key = None;
+        loop {
+            let (keys, next) = self
+                .scan_keys(&mut ctx, safe_point, next_key, GC_BATCH_SIZE)
+                .map_err(|e| {
+                    warn!("scan_keys failed on region {}: {:?}", safe_point, &e);
+                    e
+                })?;
+            if keys.is_empty() {
+                break;
+            }
+
+            next_key = self
+                .gc_keys(&mut ctx, safe_point, keys, next)
+                .map_err(|e| {
+                    warn!("gc_keys failed on region {}: {:?}", safe_point, &e);
+                    e
+                })?;
+            if next_key.is_none() {
+                break;
+            }
+        }
+
+        debug!(
+            "gc on region {}, safe_point {} has finished",
+            ctx.get_region_id(),
+            safe_point
+        );
+        Ok(())
+    }
+}
+
+impl<E: Engine> Runnable<GCTask> for GCRunner<E> {
+    fn run(&mut self, task: GCTask) {
+        GC_GCTASK_COUNTER.inc();
+
+        let timer = Instant::now();
+        let result = self.gc(task.ctx, task.safe_point);
+        GC_DURATION_HISTOGRAM.observe(duration_to_sec(timer.elapsed()));
+
+        if result.is_err() {
+            GC_GCTASK_FAIL_COUNTER.inc();
+        }
+        (task.callback)(result);
+    }
+
+    fn run_batch(&mut self, tasks: &mut Vec<GCTask>) {
+        for task in tasks.drain(..) {
+            // Do not print slow log
```

Comment:
What does this comment mean?

Reply:
The default implementation of `run_batch` records a warning log when it takes more than 1 second to handle a task. It's not proper here, so I overridden it and added a comment to state why I override it...